alf/algorithms/generator.py

```diff
@@ -119,43 +119,100 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
+                 input_dim,
+                 output_dim,
+                 hidden_size=100,
+                 num_hidden_layers=1,
                  optimizer=None,
                  name="InverseMVPAlgorithm"):
         r"""Create a InverseMVPAlgorithm.
         Args:
-            net (Network): network for predicting outputs from inputs.
+            input_dim (int): dimension of input z
+            output_dim (int): output dimension, i.e., dimension of the mvp 
+            hidden_size (int): width of hidden layers
+            num_hidden_layers (int): number of hidden layers after 
             optimizer (torch.optim.Optimizer): (optional) optimizer for training.
             name (str): name of this Algorithm.
         """
+        assert input_dim <= output_dim
+
         if optimizer is None:
             optimizer = alf.optimizers.Adam(lr=1e-3)
         super().__init__(train_state_spec=(), optimizer=optimizer, name=name)
 
-        self._net = net
+        kernel_initializer = functools.partial(
+            alf.initializers.variance_scaling_init,
+            gain=1.0 / 2.0,
+            mode='fan_in',
+            distribution='truncated_normal',
+            nonlinearity=math_ops.identity)
+
+        self._z_dim = input_dim
+        self._vec_dim = output_dim
+        z_spec = TensorSpec(shape=(self._z_dim, ))
+        vec_spec = TensorSpec(shape=(self._vec_dim, ))
+
+        self._net = EncodingNetwork(
+            (z_spec, vec_spec),
+            input_preprocessors=(torch.nn.Linear(self._z_dim, hidden_size),
+                                 torch.nn.Linear(self._vec_dim, hidden_size)),
+            preprocessing_combiner=alf.layers.NestConcat(),
+            fc_layer_params=(2 * hidden_size, ) * num_hidden_layers,
+            activation=torch.relu_,
+            kernel_initializer=kernel_initializer,
+            last_layer_size=output_dim,
+            last_activation=math_ops.identity,
+            name='InverseMVPNetwork')
 
     def predict_step(self, inputs, state=None):
         """Predict for one step of inputs.
         Args:
-            inputs (tuple of Tensors): inputs (z, eps) for prediction.
+            inputs (tuple of Tensors): inputs (z, vec) for prediction.
+            - z (Tensor): of size [N2, K] or [N2, D], representing :math:`z'`
+            - vec (Tensor): of size [N2, D] or [N2, N, D], representing 
+                :math:`\nabla_{z'}k(z', z)`.
             state: not used.
             
         Returns:
             AlgStep:
             - output (torch.Tensor): predictions of InverseMVP network
             - state: not used.
         """
-        outputs = self._net(inputs)[0]
+        z_inputs, vec = inputs
+        assert z_inputs.ndim == 2 and z_inputs.shape[-1] >= self._z_dim
+        assert vec.shape[-1] == self._vec_dim
+        assert z_inputs.shape[0] == vec.shape[0]
+
+        if z_inputs.shape[-1] > self._z_dim:
+            z_inputs = z_inputs[:, self._z_dim]  # [N2, K]
```

Comment:
```python
z_inputs = z_inputs[:, :self._z_dim]
```
I wonder how the unit test can still pass with this bug.

Reply:
Oops...updated.
It ends up no tests ever getting into this case. 
- inverse_mvp_algorithm_test has K=2, D=3, but sends in z_inputs of shape [5, 2]
- generator tests only test for K=D=2, I will explain to you the reason we did not include K<D for generator tests.